webview-cocoa.c

```diff
@@ -122,64 +133,103 @@ static void run_save_panel(id self, SEL cmd, id download, id filename,
+  ((void(*)(id, SEL))objc_msgSend)(alert, sel_registerName("release"));
 }
 
 static void run_alert_panel(id self, SEL cmd, id webView, id message, id frame,
                             void (^completionHandler)(void)) {
-  id alert =
-      objc_msgSend((id)objc_getClass("NSAlert"), sel_registerName("new"));
-  objc_msgSend(alert, sel_registerName("setIcon:"),
-               objc_msgSend((id)objc_getClass("NSImage"),
+  id alert = ((id(*)(id, SEL))objc_msgSend)((id)objc_getClass("NSAlert"), sel_registerName("new"));
+
+  ((void(*)(id, SEL, id))objc_msgSend)(alert, sel_registerName("setIcon:"),
+               ((id(*)(id, SEL, id))objc_msgSend)((id)objc_getClass("NSImage"),
                             sel_registerName("imageNamed:"),
                             get_nsstring("NSCaution")));
-  objc_msgSend(alert, sel_registerName("setShowsHelp:"), 0);
-  objc_msgSend(alert, sel_registerName("setInformativeText:"), message);
-  objc_msgSend(alert, sel_registerName("addButtonWithTitle:"),
+
+  ((void(*)(id, SEL, int))objc_msgSend)(alert, sel_registerName("setShowsHelp:"), 0);
+  ((void(*)(id, SEL, id))objc_msgSend)(alert, sel_registerName("setInformativeText:"), message);
+  ((void(*)(id, SEL, id))objc_msgSend)(alert, sel_registerName("addButtonWithTitle:"),
                get_nsstring("OK"));
-  objc_msgSend(alert, sel_registerName("runModal"));
-  objc_msgSend(alert, sel_registerName("release"));
+  ((void(*)(id, SEL))objc_msgSend)(alert, sel_registerName("runModal"));
+  ((void(*)(id, SEL))objc_msgSend)(alert, sel_registerName("release"));
   completionHandler();
 }
 
 static void download_failed(id self, SEL cmd, id download, id error) {
   printf("%s",
-         (const char *)objc_msgSend(
-             objc_msgSend(error, sel_registerName("localizedDescription")),
+         ((const char *(*)(id, SEL))objc_msgSend)(
+             ((id(*)(id, SEL))objc_msgSend)(error, sel_registerName("localizedDescription")),
              sel_registerName("UTF8String")));
 }
 
 static void make_nav_policy_decision(id self, SEL cmd, id webView, id response,
                                      void (^decisionHandler)(int)) {
-  if (objc_msgSend(response, sel_registerName("canShowMIMEType")) == 0) {
+  if (((id(*)(id, SEL))objc_msgSend)(response, sel_registerName("canShowMIMEType")) == 0) {
     decisionHandler(WKNavigationActionPolicyDownload);
   } else {
     decisionHandler(WKNavigationResponsePolicyAllow);
   }
 }
 
+
+static const char *parse_HTML_content(const char *html, int *comma_index) {
+    if (html == NULL || *html == '\0' || comma_index == NULL) {
+        return NULL; // Handling invalid input
+    }
+
+    const char *p = html;
+    char result[256]; // Assumed maximum length of the result
+
+    // Ignore whitespace at the beginning of the string
+    while (isspace(*p)) {
+        p++;
+    }
+
+    int i = 0;
+    while (*p != ',' && *p != '\0' && i < 255) {
+        if (!isspace(*p)) {

```

Comment:
Is there a reason to ignore spaces and to accept missing comma?
I would prefer to be as much as possible restrictive in parsing and to not change the URI content.
I propose something like the following:
```c
static const char *parse_data_URI(const char *uri, int *comma_index) {
  if (uri == NULL || comma_index == NULL || strncmp(uri, "data:", 5) != 0) {
    return NULL;
  }
  const char *pstart = uri + 5;
  const char *pend = uri + 255;
  const char *p = pstart;
  while (*p != ',') {
    if (*p <= ' ' || *p >= '\x7f' || p > pend) {
      return NULL;
    }
    p++;
  }
  int size = (int)(p - pstart);
  char *result = malloc(size + 1);
  if (result != NULL) {
    strncpy(result, pstart, size);
    result[size] = '\0';
    *comma_index = (int)(p - uri);
  }
  return result;
}
```